toolkit/tools/pkg/imagecustomizerlib/customizeuki.go

```diff
@@ -0,0 +1,538 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"encoding/json"
+	"fmt"
+	"gopkg.in/ini.v1"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/azurelinux/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/file"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safeloopback"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safemount"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+)
+
+const (
+	BootDir           = "boot"
+	GrubCfgDir        = "grub2/grub.cfg"
+	KernelCmdlineArgs = "kernel-cmdline-args"
+	KernelPrefix      = "vmlinuz-"
+	UkiBuildDir       = "UkiBuildDir"
+	UkiOutputDir      = "EFI/Linux"
+)
+
+func prepareUki(buildDir string, uki *imagecustomizerapi.Uki, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	if uki == nil {
+		return nil
+	}
+
+	logger.Log.Infof("Enabling UKI")
+
+	// Check UKI dependency packages.
+	err = validateUkiDependencies(imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to validate package dependencies for uki:\n%w", err)
+	}
+
+	// Create necessary directories for UKI.
+	err = createUkiDirectories(buildDir, imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to create UKI directories:\n%w", err)
+	}
+
+	// This code installs the systemd-boot bootloader into the EFI system partition (ESP).
+	// Note: When proper support for systemd-boot is implemented, the `bootctl install` command
+	// will likely be invoked as part of the `hardResetBootLoader()` function under BootLoader structure.
+	//
+	// The command being executed is:
+	//     bootctl install --no-variables
+	// This performs the following steps:
+	//   1. Creates the necessary directories in the ESP, such as:
+	//        - /boot/efi/EFI/systemd
+	//        - /boot/efi/loader
+	//        - /boot/efi/loader/entries
+	//   2. Copies the systemd bootloader binary from the host filesystem to the ESP:
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/systemd/systemd-bootx64.efi
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/BOOT/BOOTX64.EFI
```

Comment:
Are you sure this is accurate? From my testing it copies shim to `/boot/efi/EFI/BOOT/bootx64.efi` and copies systemd-boot to `/boot/efi/EFI/BOOT/grubx64.efi`

(Yes, **grub**x64.efi. The shim binary is [hardcoded ](https://github.com/rhboot/shim/blob/3cf0e099b4066d6615ffa7d11e5b62428dd1c4d5/shim.h#L73)to look for that filename)

Reply:
We'll probably have to investigate further for secure boot, but doesn't have to be part of this PR